File: solutions/day-1/python/day1.py
from pathlib import Path
DIALLEN = 100


def parse_rotation(s):
    return (-1 if s[0] == "L" else 1) * int(s[1:])


def part1(lines):
    position, password = 50, 0
    for rotation in map(parse_rotation, lines):
        position = (position + rotation) % DIALLEN
        password += (position == 0)
    return f"Door password: {password}"


def part2(lines):
    position, password = 50, 0
    for rotation in map(parse_rotation, lines):
        password += (position != 0) - int((position + rotation) /
                                          DIALLEN) if position + rotation <= 0 else (position + rotation) // DIALLEN
        position = (position + rotation) % DIALLEN
    return f"Door password using password method 0x434C49434B: {password}"


file_name = "day1.in"

lines = Path(file_name).read_text().splitlines()
print(part1(lines))
print(part2(lines))

File: solutions/day-1/rust/day1.rs
use std::fs;

fn parse_rotation(rotation: &str) -> i32 {
    (if rotation.starts_with("L") { -1 } else { 1 })
        * rotation[1..]
            .parse::<i32>()
            .expect("Failed to parse rotation")
}

const DIAL_LEN: i32 = 100;

fn part1(rotations: &Vec<&str>) -> i32 {
    let (_final_position, password) = rotations.iter().map(|&r| parse_rotation(r)).fold(
        (50, 0),
        |(position, password), rotation| {
            let next_position = (position + rotation).rem_euclid(DIAL_LEN);
            (next_position, password + (next_position == 0) as i32)
        },
    );
    password
}

fn part2(rotations: &Vec<&str>) -> i32 {
    let (_final_position, password) = rotations.iter().map(|&r| parse_rotation(r)).fold(
        (50, 0),
        |(position, password), rotation| {
            let shifted_pos = position + rotation;
            (
                shifted_pos.rem_euclid(DIAL_LEN),
                password
                    + if shifted_pos <= 0 {
                        (position != 0) as i32 - shifted_pos / DIAL_LEN
                    } else {
                        shifted_pos / DIAL_LEN
                    },
            )
        },
    );
    password
}

fn main() {
    let file_name = "day1.in";
    let content = fs::read_to_string(file_name).expect("Failed to read input file");
    let rotations: Vec<&str> = content.lines().collect();
    println!("Part 1: {}", part1(&rotations));
    println!("Part 2: {}", part2(&rotations));
}
